filters/HAGFilter.cpp

```diff
@@ -45,6 +45,133 @@
+        double lambda2 = ((y3-y1) * (x-x3) + (x1-x3) * (y-y3)) / detT;
+        if (lambda1 >= 0 && lambda1 <= 1 && lambda2 >= 0 && lambda2 <= 1)
+        {
+            double sum = lambda1 + lambda2;
+            if (sum <= 1)
+            {
+                double lambda3 = 1 - sum;
+                z = (lambda1 * z1) + (lambda2 * z2) + (lambda3 * z3);
+            }
+        }
+    }
+    return z;
+}
+
+
+// The non-ground point (x0, y0) is in exactly 0 or 1 of the triangles of
+// the ground triangulation, so when we find a triangle containing the point,
+// return the interpolated z.
+// (I suppose the point could be on a edge of two triangles, but the
+//  result is the same, so this is still good.)
+double delaunay_interp_ground(double x0, double y0, PointViewPtr gView,
+    const PointIdList& ids)
+{
+    using namespace pdal::Dimension;
+
+    // Delaunay-based interpolation
+    std::vector<double> neighbors;
+
+    for (size_t j = 0; j < ids.size(); ++j)
+    {
+        neighbors.push_back(gView->getFieldAs<double>(Id::X, ids[j]));
+        neighbors.push_back(gView->getFieldAs<double>(Id::Y, ids[j]));
+    }
+
+    delaunator::Delaunator triangulation(neighbors);
+    const std::vector<size_t>& triangles(triangulation.triangles);
+
+    for (size_t j = 0; j < triangles.size(); j += 3)
+    {
+        auto ai = triangles[j+0];
+        auto bi = triangles[j+1];
+        auto ci = triangles[j+2];
+        double ax = gView->getFieldAs<double>(Id::X, ids[ai]);
+        double ay = gView->getFieldAs<double>(Id::Y, ids[ai]);
+        double az = gView->getFieldAs<double>(Id::Z, ids[ai]);
+
+        double bx = gView->getFieldAs<double>(Id::X, ids[bi]);
+        double by = gView->getFieldAs<double>(Id::Y, ids[bi]);
+        double bz = gView->getFieldAs<double>(Id::Z, ids[bi]);
+
+        double cx = gView->getFieldAs<double>(Id::X, ids[ci]);
+        double cy = gView->getFieldAs<double>(Id::Y, ids[ci]);
+        double cz = gView->getFieldAs<double>(Id::Z, ids[ci]);
+
+        // Returns infinity unless the point x0/y0 is in the triangle.
+        double z1 = distance_along_z(ax, ay, az, bx, by, bz,
+                cx, cy, cz, x0, y0);
+        if (z1 != std::numeric_limits<double>::infinity())
+            return z1;
+    }
+    // If the non ground point was outside the triangulation of ground
+    // points, just use the Z coordinate of the closest
+    // ground point.
+    return gView->getFieldAs<double>(Id::Z, ids[0]);
+}
+
+double neighbor_interp_ground(PointViewPtr gView, const PointIdList& ids,
+    const std::vector<double>& sqr_dists, double maxDistance2, double zDefault)
+{
+    double weights = 0;
+    double z_accumulator = 0;
+
+    for (size_t j = 0; j < ids.size(); ++j)
+    {
+        auto z = gView->getFieldAs<double>(Dimension::Id::Z, ids[j]);
+        double sqr_dist = sqr_dists[j];
+        if (maxDistance2 > 0 && sqr_dist > maxDistance2)
+            break;
```

Comment:
If we are going to keep the current behavior, which is to not give a result at all (well, give the default Z that is) if any of the neighbors exceeds the max distance, then we can just compare the last element of `sqr_dist` before we even enter the for loop and save a few iterations here and there.

Reply:
This isn't what is happening.  We use all the neighbors whose distance from the test point is less than the max distance that was specified.

Reply:
Right, I got it confused again. Sorry. I'll take another look at the documentation as I'm still not convinced it reflects what's being done here.